src/Microsoft.AspNet.Http.Features/FeatureCollection.cs

```diff
@@ -78,89 +64,39 @@ IEnumerator IEnumerable.GetEnumerator()
             return GetEnumerator();
         }
 
-        public void Add(KeyValuePair<Type, object> item)
-        {
-            SetInterface(item.Key, item.Value);
-        }
-
-        public void Clear()
-        {
-            throw new NotImplementedException();
-        }
-
-        public bool Contains(KeyValuePair<Type, object> item)
-        {
-            object value;
-            return TryGetValue(item.Key, out value) && Equals(item.Value, value);
-        }
-
-        public void CopyTo(KeyValuePair<Type, object>[] array, int arrayIndex)
-        {
-            throw new NotImplementedException();
-        }
-
-        public bool Remove(KeyValuePair<Type, object> item)
-        {
-            return Contains(item) && Remove(item.Key);
-        }
-
-        public int Count
-        {
-            get { throw new NotImplementedException(); }
-        }
-
-        public bool IsReadOnly
-        {
-            get { return false; }
-        }
-
-        public bool ContainsKey([NotNull] Type key)
-        {
-            return GetInterface(key) != null;
-        }
-
-        public void Add([NotNull] Type key, [NotNull] object value)
+        public IEnumerator<KeyValuePair<Type, object>> GetEnumerator()
         {
-            if (ContainsKey(key))
+            foreach (var pair in _featureByFeatureType)
             {
-                throw new ArgumentException();
+                yield return pair;
             }
-            SetInterface(key, value);
-        }
 
-        public bool Remove([NotNull] Type key)
-        {
-            lock (_containerSync)
+            if (_defaults != null)
             {
-                if (_featureByFeatureType.Remove(key))
+                // Don't return features masked by the wrapper.
+                foreach (var pair in _defaults.Except(_featureByFeatureType, FeatureKeyComparer))
                 {
-                    _containerRevision++;
-                    return true;
+                    yield return pair;
                 }
-                return false;
             }
         }
 
-        public bool TryGetValue([NotNull] Type key, out object value)
-        {
-            value = GetInterface(key);
-            return value != null;
-        }
-
-        public object this[Type key]
+        public virtual void Dispose()
         {
-            get { return GetInterface(key); }
-            set { SetInterface(key, value); }
+            _defaults?.Dispose();
         }
 
-        public ICollection<Type> Keys
+        private class KeyComparer : IEqualityComparer<KeyValuePair<Type, object>>
         {
-            get { throw new NotImplementedException(); }
-        }
+            public bool Equals(KeyValuePair<Type, object> x, KeyValuePair<Type, object> y)
+            {
+                return x.Key.Equals(y.Key);
```

Comment:
This is only used on line 77 to filter out features that exist in both collections, so we only want to match by key.